service/cat/cat_service_impl.go

```diff
@@ -0,0 +1,71 @@
+package cat_service
+
+import (
+	cat_entity "cats-social/entity/cat"
+	exc "cats-social/exceptions"
+	catRep "cats-social/repository/cat"
+	authService "cats-social/service/auth"
+	"fmt"
+	"strings"
+
+	"github.com/go-playground/validator"
+	"github.com/gofiber/fiber/v2"
+	"github.com/jackc/pgx/v5/pgxpool"
+)
+
+type CatServiceImpl struct {
+	CatRepository catRep.CatRepository
+	DBPool        *pgxpool.Pool
+	AuthService   authService.AuthService
+	Validator     *validator.Validate
+}
+
+func NewUserService(catRepository catRep.CatRepository, dbPool *pgxpool.Pool, authService authService.AuthService, validator *validator.Validate) CatService {
+	return &CatServiceImpl{
+		CatRepository: catRepository,
+		DBPool:        dbPool,
+		AuthService:   authService,
+		Validator:     validator,
+	}
+}
+
+func (service *CatServiceImpl) Create(ctx *fiber.Ctx, req cat_entity.CatCreateRequest) (cat_entity.CatCreateResponse, error) {
+	if err := service.Validator.Struct(req); err != nil {
+		return cat_entity.CatCreateResponse{}, exc.BadRequestException(fmt.Sprintf("%s", err))
+	}
+
+	tx, err := service.DBPool.Begin(ctx.UserContext())
+	if err != nil {
+		return cat_entity.CatCreateResponse{}, exc.InternalServerException(fmt.Sprintf("Internal Server Error: %s", err))
+	}
+	defer tx.Rollback(ctx.UserContext())
+
+	userId, err := authService.NewAuthService().GetValidUser(ctx)
+	if err != nil {
+		return cat_entity.CatCreateResponse{}, exc.UnauthorizedException("Unauthorized")
+	}
+	cat := cat_entity.Cat{
+		Name:        req.Name,
+		Race:        req.Race,
+		Sex:         req.Sex,
+		AgeInMonth:  req.AgeInMonth,
+		Description: req.Description,
+		ImageURLs:   req.ImageURLs,
+	}
+
+	catRegistered, err := catRep.NewCatRepository().Create(ctx.UserContext(), tx, cat, userId)
+	if err != nil {
+		if strings.Contains(err.Error(), "no rows in result set") {
```

Comment:
ini error `no rows in result set` case nya hanya ketika user id nya invalid?

Reply:
iya mas betul, jadi create cat-nya ke cancel gitu, karena jwt-nya masih valid, jadi userId-nya tetep didapetin walaupun ga valid gitu, akhirnya data-nya tetep tersimpan

Reply:
hmm aneh juga padahal dah gw defer Rollback diatas

Reply:
trus kenapa bisa invalid user id nya pdhal jwt nya masih live?

Reply:
masuk mas, soalnya userId-nya dr jwt kan masih valid walaupun userId-nya udah kehapus/ ga ada. Jadinya aku bikin query yg ngecheck dulu kalo userId-nya tuh ada ga? nah nanti kalau ga ada keluar error yg `no rows in result set` itu dan gabakal disimpan gitu. coba check di `cat_repository_impl`.

Reply:
oalaaah noted2 baru liat gw query nya oke2 mb, jadi checking user dlu pake subquery ya sblm insert cat. nicee

Reply:
most likely jarang sih yaa, paling kalo ada org yg bisa dapetin jwt tokennya nya aja wwkwk tp ya jarang la

Reply:
oke oke mas